### ISSUE TITLE
Add support for subscriptions not bound to message version

### ISSUE THREAD

**Issue by mauroservienti** [Particular/NServiceBus.RavenDB#319]
As of now RavenDB Persistence supports only subscriptions bound to message versioning ([doco page](https://docs.particular.net/nservicebus/ravendb/subscription-versioning)). The main reason is due to the fact that we originally stored the subscription document including in the document ID the version as well:

```
// Pseudo-code only
subscriptionDocumentId = $"Subscriptions/{HashOf(messageType.Name + messageType.AssemblyVersion.Major)}";
```

Changing the above is a complex exercise that is basically impossible to solve in a no downtime fashion and is also a huge breaking change making it impossible for previous versions to deal with the new subscription format.

There is although an option that we might not have considered:

* at endpoint startup time we could load all subscriptions
* iterate over them
* create a new Subscription document that uses the sole message type name for the ID
* add to this new subscription document only a reference to the existing one
* New versions of the Persistence at publish time will query the storage for the new subscription document "including" the referenced one

The above would allow us to go on storing subscriptions in one single place, the old subscription format, meaning that 2 different versions of the RavenDB Persistence will have a single source of truth, and only newer versions will use the new document point to load subscribers.

There is only one single not supported scenario, that is when a new subscription request comes in from a new endpoint (never seen before) for an event never published before. When this happens and the subscription request is handled by the old version endpoint (in a scaled out scenario) the pointer document is not created meaning that new versions won't know about the subscriptions until the endpoint instance is restarted. We could alleviate this by polling the storage for changes and ensure that they are in synch.

We could also introduce a setting that enables the above behavior requiring users to opt-in, and another setting that says "use the new format only, I know what I'm doing", setting that could be the default behavior in the next major release.

@Particular/ravendb-persistence-maintainers thoughts?

## Plan of Attack

* [x] Implement support for non-versioned subscriptions #327 
* [x] document that the new feature is https://github.com/Particular/docs.particular.net/pull/2867
   * opt-in
   * non-backward compatible with the way actual subscription format
* [x] release https://github.com/Particular/NServiceBus.RavenDB/pull/332

**Comment by marktaling** [comment]
For us this is a real issue. We version our assemblies, resulting in unexpected runtime behavior if we do not update all our services simultaneously.

This solution seems rather elegant to me. We came to something similar in case we would have to fix it ourselves...

Personally I do not really have a problem with requiring a one-time downtime per service to upgrade to this scenario, if it results in the issue being resolved. (Then again, I do work in a business where we are allowed some occasional maintenance downtime)

**Comment by mauroservienti** [comment]
We had an internal discussion and we see another, simpler, option.

### Upgrade with downtime required

Introduce a new setting, such as `DoNotUseVersionedSubscriptions()` or similar, that makes the endpoint look ONLY for subscription documents without version.
In this case each endpoint, if scaled out, requires a downtime to be upgraded, because it won't be backward compatible when the setting is on.

### Migration

A migration process needs to be put in place, and we have 2 options:

- do not migrate anything, but require subscribers to be restarted, at restart time new subscriptions requests will be sent out and the publisher will be up-to-date after awhile
- provide a small migration script/tool to run once before starting endpoints after the upgrade

I personally prefer the latter.

### Last but not least: evolution

We can introduce the setting in a minor release of the actual supported packages. Once we have that:

- in the next major (e.g. NServiceBus.ravenDB V5) we can invert the behavior making unversioned subscriptions the default and an opt-in setting such as `DoUseVersionedSubscriptions()` to go on using the versioned behavior.
- At the same time we can obsolete with a warning the  `DoNotUseVersionedSubscriptions()` to make users aware that that is the new default behavior
- in the major after (e.g. NServiceBus.ravenDB V6) we can start obsoleting versioned subscriptions by obsolete with a warning `DoUseVersionedSubscriptions()` and with a error `DoNotUseVersionedSubscriptions()` that is the default behavior
- in the major after (e.g. NServiceBus.ravenDB V7) `DoNotUseVersionedSubscriptions()` will be removed and subscriptions will be supported as non versioned only

We'll have another internal brainstorming session on the above plan to see if we're missing something.

**Comment by mauroservienti** [comment]
Raised https://github.com/Particular/NServiceBus.RavenDB/pull/325 to start working on this.

**Comment by mauroservienti** [comment]
Added PoA

**Comment by mauroservienti** [comment]
Doco PR (draft) created https://github.com/Particular/docs.particular.net/pull/2867

**Comment by mauroservienti** [comment]
I'd leave this open till doco PR is merged and package released

**Comment by mauroservienti** [comment]
@marktaling we're in the process of releasing `4.2.0` with the support for unversioned subscriptions.
I just promoted `4.2.0-alpha33` on MyGet, would be awesome if you can verify that it solves the issue you're facing in your environment.

In order to add the Myget feed you can point your Visual Studio to https://www.myget.org/F/particular/api/v3/index.json or run the following a the Package Manager console:

```
Install-Package NServiceBus.RavenDB -Version 4.2.0-alpha0033 -Source https://www.myget.org/F/particular/api/v3/index.json
```

**Comment by marktaling** [comment]
Sadly I am unable to do so at the moment. We ran accross this in NServiceBus 4. As I was working on the upgrade to NServiceBus 6.x I sent in this call. Currently the upgrade is on hold do to a staffing issue.... I won't be continueing the upgrade for a few weeks yet...

**Comment by mauroservienti** [comment]
No problems @marktaling, we'll go ahead and release. It should be released by end of this week. Thanks for chiming in.

**Comment by mauroservienti** [comment]
closing this one as we are in the process of releasing the new 4.2.0 version.
